src/Compilers/CSharp/Test/Emit3/Semantics/PatternMatchingTests4.cs

```diff
@@ -4829,5 +4842,4429 @@ class D : A { }
+    public int P4 { get; set; }
+}
+""";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75506")]
+        public void RedundantPattern_RecursivePattern_InputTypeOfNull()
+        {
+            var source = """
+S s = default;
+_ = s is { P: string or null }; // 1
+_ = s is { P: string and null }; // 2
+_ = s is { P: null or "" }; 
+
+C c = null;
+_ = c is { P: string or null }; // 3
+_ = c is { P: string and null }; // 4
+_ = c is { P: null or "" };
+_ = c is not null and { P: string or null }; // 5
+
+object o = null;
+_ = o is S { P: string or null }; // 6
+_ = o is C { P: string or null }; // 7
+
+public struct S
+{
+    public string P { get; set; }
+}
+
+public class C
+{
+    public string P { get; set; }
+}
+""";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (2,15): hidden CS9274: The pattern is redundant.
+                // _ = s is { P: string or null }; // 1
+                Diagnostic(ErrorCode.HDN_RedundantPattern, "string or null").WithLocation(2, 15),
+                // (3,5): error CS8518: An expression of type 'S' can never match the provided pattern.
+                // _ = s is { P: string and null }; // 2
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "s is { P: string and null }").WithArguments("S").WithLocation(3, 5),
+                // (7,15): hidden CS9274: The pattern is redundant.
+                // _ = c is { P: string or null }; // 3
+                Diagnostic(ErrorCode.HDN_RedundantPattern, "string or null").WithLocation(7, 15),
+                // (8,5): error CS8518: An expression of type 'C' can never match the provided pattern.
+                // _ = c is { P: string and null }; // 4
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "c is { P: string and null }").WithArguments("C").WithLocation(8, 5),
+                // (10,28): warning CS9275: The pattern is redundant.
+                // _ = c is not null and { P: string or null }; // 5
+                Diagnostic(ErrorCode.WRN_RedundantPattern, "string or null").WithLocation(10, 28),
+                // (13,17): hidden CS9274: The pattern is redundant.
+                // _ = o is S { P: string or null }; // 6
+                Diagnostic(ErrorCode.HDN_RedundantPattern, "string or null").WithLocation(13, 17),
+                // (14,17): hidden CS9274: The pattern is redundant.
+                // _ = o is C { P: string or null }; // 7
+                Diagnostic(ErrorCode.HDN_RedundantPattern, "string or null").WithLocation(14, 17));
+        }
+
+        [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/75506")]
+        public void RedundantPattern_RecursivePattern_OneAndNotOne()
+        {
+            var source = """
+var c = new C();
+
+_ =  c is ({F: 1 and not 1} and {O: not A or B}) or {F: 2}; // 1
```

Comment:
Could you please explain a bit why the part we blame here is the entire `not A or B`? `B` itself seems redundant, but `not A` seems meaningful. It's just that it's preceded by a "contradiction" pattern.

Reply:
I'm not sure. Sorry, I didn't step through to understand how this falls out of the algorithm. Let me know if you still think it's worthwhile